### ISSUE TITLE
Fisherman should provide an official way for plugins to register keybindings

### ISSUE THREAD

**Issue by kballard** [fisherman/fisherman#42]
Fish currently invokes the user function `fish_user_key_bindings` whenever the `fish_key_bindings` variable changes (and prior to the first prompt being displayed). That's the official customization point for loading in your own keybindings. Anything set manually with `bind` will not survive the session, and will also get erased if the user changes the `fish_key_bindings` variable.

Unfortunately, this isn't very amenable to a plugin infrastructure, because multiple plugins that want to set keybindings would overwrite each other when setting this function (and would overwrite the user's own keybindings). You can see this in the OMF plugin [oh-my-fish/plugin-bang-bang](https://github.com/oh-my-fish/plugin-bang-bang) which simply provides its own `fish_user_key_bindings` function.

I think Fisherman can do better.

I think the right approach here is for Fisherman to define its own `fish_user_key_bindings` function. Since Fisherman puts itself in the `fish_function_path` prior to `$XDG_CONFIG_HOME/fish/functions`, Fisherman's version will override the user's. Fisherman can also alter the plugin structure to allow for a file named `key_bindings.fish`, which will get copied into a new folder `$fisher_config/keybindings` when the plugin is installed. Fisherman's `fish_user_key_bindings` can then source every file in this folder.

Furthermore, Fisherman's version of `functions/fish_user_key_bindings.fish` can explicitly check the remainder of the `$fish_function_path` list (everything after the fisherman path) and look for another file called `fish_user_key_bindings.fish`; if it finds one, it can source it, then use `functions -c` to copy the other `fish_user_key_bindings` to some other name. This way it can execute that function at the end of its own `fish_user_key_bindings` (so user keybindings will override plugin ones). This way users can still define their own `fish_user_key_bindings` functions and it will still be invoked as appropriate; the only real problem is if they modify the file, re-running `fish_user_key_bindings` won't reapply it. A possible workaround there is to actually search for the file inside the invocation of `fish_user_key_bindings`, but then you have to do weird things like rename Fisherman's version of the function, execute the file, check to make sure it actually defined its own `fish_user_key_bindings` function, execute the function, and then rename Fisherman's copy back to the normal name. This sounds doable, but I'm not sure if there's any potential issues there.

I'm actually kind of surprised OMF doesn't already do something like this.

**Comment by kballard** [comment]
An alternative approach would be to just define something like `fisher --evaluate-key-bindings` and have the Fisherman installation process modify the user's `fish_user_key_bindings` to insert that call at the top (or to define the function if it's not defined). Though this has the potential issue where you'd have to parse the file sufficiently to actually find the function declaration (there's no guarantee it's at the top of the file) and detect the first line of the body.

**Comment by bucaran** [comment]
To be fair I think the OMF folks are working on something to do with key bindings, but I am not sure. 

In any case, yes, this is a must-have for 1.0. I'll add this to 0.5.0

**Comment by bucaran** [comment]
@kballard I finally had time to go over this. 

Copying the function or even doing the simplest tricks inside `fish_user_key_bindings` slows down shell, so I had to rule that out.

The good news is I came up with another solution taking some cues from your suggestion that has the following advantages:

+ Does not slow down shell start.
+ Does not require Fisherman to provide his own `fish_user_key_bindings`.
+ Honors any previously existing user key bindings.
+ Allows plugin to define their own key bindings and coexist with the user's key bindings.
+ If the user updates his `fish_user_key_bindings`, re-running the function **does** update the key bidings.

I'll explain how this was achieved in the CHANGELOG when 0.5.0 ships.
